elastic-buffer-sim/exe/Main.hs

```diff
@@ -2,70 +2,644 @@
+      Diamond       -> (,dName) <$> plotDiamond settings
+      Line n        -> (,dName) <$> plotLine settings n
+      HyperCube n   -> (,dName) <$> plotHyperCube settings n
+      Grid r c      -> (,dName) <$> plotGrid settings r c
+      Torus2D r c   -> (,dName) <$> plotTorus2D settings r c
+      Torus3D r c p -> (,dName) <$> plotTorus3D settings r c p
+      Tree d c      -> (,dName) <$> plotTree settings d c
+      Star n        -> (,dName) <$> plotStar settings n
+      Cycle n       -> (,dName) <$> plotCyclic settings n
+      Complete n    -> (,dName) <$> plotComplete settings n
+      Random n      -> (,dName) <$> (randomGraph n >>= plotGraph settings)
+      DotFile f     -> (fromDot <$> readFile f) >>= \case
+        Right (g, name) -> (,name) <$> plotGraph settings g
+        Left err        -> die $ "ERROR: Invalid DOT file - " <> f <> "\n" <> err
+    Nothing ->
+      handleParseResult $ Failure
+        $ parserFailure defaultPrefs cliParser (ShowHelpText Nothing) []
+
+  let topologyFile = outDir </> "topology.gv"
+  writeFile topologyFile $ (<> "\n") $ render $ toDot g name
+  BS.writeFile (outDir </> "simulate.json") $ encode options
+     { stable   = isStable
+     , offsets  = offs
+     , jsonArgs = Nothing
+     , topology = case topology of
+         Just (Random _) -> Just $ DotFile topologyFile
+         _               -> topology
+     }
+
+  if isStable
+  then exitSuccess
+  else exitFailure
+
+randomGraph :: Int -> IO Graph
+randomGraph n = do
+  let is = [0,1..n - 1]
+
+  -- get some random vertex permuation for ensuring connectivity
+  aP <- newListArray (0, n - 1) is
+  replicateM_ (10 * n) $ do
+    p1 <- randomRIO (0, n - 1)
+    p2 <- randomRIO (0, n - 1)
+    v1 <- readArray aP p1
+    v2 <- readArray aP p2
+    writeArray aP p1 v2
+    writeArray aP p2 v1
+
+  randomPermutation <- getElems (aP :: IOUArray Int Int)
+
+  -- get some random edge availability matrix for self-loop-free,
+  -- undirected graphs
+  aE <- replicateM (n * n) randomIO >>= newListArray ((0, 0), (n - 1, n - 1))
+  forM_ is $ \i ->
+    writeArray aE (i, i) False
+  forM_ [(i, j) | i <- is, j <- is, i /= j] $ \(i, j) -> do
+    d1 <- readArray aE (i, j)
+    d2 <- readArray aE (j, i)
+    when (d1 /= d2) $ do
+      writeArray aE (i, j) False
+      writeArray aE (j, i) False
+
+  -- ensure connectivity
+  forM_ (unchain randomPermutation) $ \(i, j) -> do
+    writeArray aE (i, j) True
+    writeArray aE (j, i) True
+
+  available <- freeze (aE :: IOUArray (Int, Int) Bool)
+
+  -- create the graph
+  return $ buildG (0, n - 1)
+    [ (i, j)
+    | i <- is
+    , j <- is
+    , available A.! (i, j)
+    ]
+
+toDot :: Graph -> String -> (Bool, GraphType, Maybe Name, [Statement])
+toDot g name = (True, Graph, Just $ StringID name, map asEdgeStatement $ edges g)
+ where
+  asEdgeStatement (x, y) =
+    EdgeStatement (map (\i -> NodeRef (XMLID ('n' : show i)) Nothing) [x,y]) []
+
+fromDot :: String -> Either String (Graph, String)
```

Comment:
Could you document the arguments/results of this function? Could you motivate its existence? I'm not sure when we would need to _read_ a dot file.

Reply:
> Could you document the arguments/results of this function?

Sure, I'll add some documentation. 

> Could you motivate its existence?

Its primarily there for reproducabilty. Each simulated topology is also written to `_build` as dot graph. Hence, we want to be able to read it back for repeating a run on exactly the same input. The `--json-args` feature also makes use of that. Note that we also support randomly generated topologies by now. Thus, if an interesting outcome is produced by that at some point, we wouldn't have any chance to reproduce that result otherwise.

Reply:
Ok!